tools/topology/src/lib.rs

```diff
@@ -0,0 +1,331 @@
+                let Some((endpoints, inverse_endpoint)) =
+                    channels.get(&network).and_then(|endpoints| {
+                        endpoints
+                            .get(host_network)
+                            .map(|&inverse_endpoint| (endpoints, inverse_endpoint))
+                    })
+                else {
+                    unreachable!("Inverse channel endpoint has to be defined!")
+                };
+
+                (
+                    network,
+                    endpoints,
+                    vec![HostToDexPathChannel {
+                        endpoint,
+                        inverse_endpoint,
+                    }],
+                )
+            })
+            .collect();
+
+        let mut traversed_networks = BTreeSet::from([host_network]);
+
+        Ok('find_connection: loop {
+            let Some((network, endpoints, mut walked_endpoints)) = endpoints_deque.pop_front()
+            else {
+                return Err(error::CurrencyDefinitions::HostNotConnectedToDex);
+            };
+
+            if network == dex_network {
+                break 'find_connection walked_endpoints;
+            }
+
+            let mut endpoints = endpoints
+                .iter()
+                .filter(|&(network, _)| traversed_networks.insert(network));
+
+            let Some((&next_network, &endpoint)) = endpoints.next() else {
+                continue;
+            };
+
+            if next_network != dex_network {
+                for (&next_network, &endpoint) in endpoints {
+                    let Some(endpoints) = channels.get(next_network) else {
+                        unreachable!()
+                    };
+
+                    let Some(&inverse_endpoint) = endpoints.get(network) else {
+                        unreachable!()
+                    };
+
+                    let host_dex_path_channel = HostToDexPathChannel {
+                        endpoint,
+                        inverse_endpoint,
+                    };
+
+                    if next_network == dex_network {
+                        walked_endpoints.push(host_dex_path_channel);
+
+                        break 'find_connection walked_endpoints;
+                    }
+
+                    let mut walked_endpoints = walked_endpoints.clone();
+
+                    walked_endpoints.push(host_dex_path_channel);
+
+                    endpoints_deque.push_back((next_network, endpoints, walked_endpoints));
+                }
+            }
+
+            let Some(endpoints) = channels.get(next_network) else {
+                unreachable!()
+            };
+
+            let Some(&inverse_endpoint) = endpoints.get(network) else {
+                unreachable!()
+            };
+
+            walked_endpoints.push(HostToDexPathChannel {
+                endpoint,
+                inverse_endpoint,
+            });
+
+            if next_network == dex_network {
+                break 'find_connection walked_endpoints;
```

Comment:
[everywhere] Please refrain from using loop labels - they make the code hard to read and indicate the presence of hidden inline functions - please refactor by extracting code into functions and employing single function return best practice.

Reply:
The reason I did it like this was that there isn't an easy way to not waste time and resources on a `.clone()` call. I'll try to tackle the problem a bit more.